src/PooledArrays.jl

```diff
@@ -330,27 +442,44 @@ Base.convert(::Type{Array}, pa::PooledArray{T, R, N}) where {T, R, N} = convert(
 ##
 ##############################################################################
 
-# Scalar case
-Base.@propagate_inbounds function Base.getindex(pa::PooledArray, I::Integer...)
-    idx = pa.refs[I...]
-    iszero(idx) && throw(UndefRefError())
-    return @inbounds pa.pool[idx]
+# We need separate functions due to dispatch ambiguities
+
+for T in (PooledArray, SubArray{<:Any, <:Any, <:PooledArray})
+    @eval Base.@propagate_inbounds function Base.getindex(A::$T, I::Integer...)
+        idx = DataAPI.refarray(A)[I...]
+        iszero(idx) && throw(UndefRefError())
+        return @inbounds DataAPI.refpool(A)[idx]
+    end
+
+    @eval Base.@propagate_inbounds function Base.getindex(A::$T, I::Union{Real, AbstractVector}...)
+        # make sure we do not increase A.refcount in case creation of newrefs fails
+        newrefs = DataAPI.refarray(A)[I...]
+        @assert newrefs isa AbstractArray
+        Threads.atomic_add!(refcount(A), 1)
+        return PooledArray(RefArray(newrefs), DataAPI.invrefpool(A), DataAPI.refpool(A), refcount(A))
+    end
 end
 
-Base.@propagate_inbounds function Base.isassigned(pa::PooledArray, I::Int...)
-    !iszero(pa.refs[I...])
+if VERSION < v"1.1"
+    Base.@propagate_inbounds function Base.getindex(A::SubArray{T,D,P,I,true} ,
+                                                    i::Int) where {I<:Tuple{Union{Base.Slice,
+                                                                                  AbstractUnitRange},
+                                                                            Vararg{Any}}, P<:PooledArray, T, D}
+        idx = DataAPI.refarray(A)[i]
+        iszero(idx) && throw(UndefRefError())
+        return @inbounds DataAPI.refpool(A)[idx]
+    end
 end
 
-# Vector case
-Base.@propagate_inbounds function Base.getindex(A::PooledArray, I::Union{Real,AbstractVector}...)
-    PooledArray(RefArray(getindex(A.refs, I...)), copy(A.invpool))
+Base.@propagate_inbounds function Base.getindex(A::SubArray{<:Any, N, <:PooledArray}, I::Vararg{Int,N}) where {T,N}
```

Comment:
```suggestion
# Defined to avoid ambiguities with Base
Base.@propagate_inbounds function Base.getindex(A::SubArray{<:Any, N, <:PooledArray}, I::Vararg{Int,N}) where {T,N}
```